### PR TITLE
Add new data type: "Fhir"

### PR DESCRIPTION
**Pull request by sowu880**
1. Add new data type "Fhir".
2. Add new FhirProcessor.
3. Update TemplateManagement for Stu3ToR4 templates.

### CODE REVIEW

File: data/Templates/Stu3ToR4/metadata.json [hunk 1]
```diff
@@ -1,3 +1,3 @@
 {
-	"type": "json"
+	"type": "fhir"
```
Comment (moria97) [inline data/Templates/Stu3ToR4/metadata.json:2]: Shall we distinguish `fhir-stu3` from `fhir-r4`? Though currently we only need r3->r4 transformation, but possibly we need r4->r5 in the future?

File: src/Microsoft.Health.Fhir.Liquid.Converter.Tool/ConverterLogicHandler.cs [hunk 2]
```diff
@@ -110,6 +111,7 @@ private static ITemplateProvider CreateTemplateProvider(DataType dataType, strin
                 DataType.Hl7v2 => new TemplateProvider(templateDirectory, DataType.Hl7v2),
                 DataType.Ccda => new TemplateProvider(templateDirectory, DataType.Ccda),
                 DataType.Json => new TemplateProvider(templateDirectory, DataType.Json),
+                DataType.Fhir => new TemplateProvider(templateDirectory, DataType.Fhir),
                 _ => throw new NotImplementedException($"The conversion from data type {dataType} to FHIR is not supported")
```
Comment (qiwjin) [inline src/Microsoft.Health.Fhir.Liquid.Converter.Tool/ConverterLogicHandler.cs:115]: maybe we can remove this switch, just use new TemplateProvider(templateDirectory, dataType)?

File: src/Microsoft.Health.Fhir.Liquid.Converter/Processors/FhirProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using Microsoft.Health.Fhir.Liquid.Converter.Models;
+
+namespace Microsoft.Health.Fhir.Liquid.Converter.Processors
+{
+    public class FhirProcessor : JsonProcessor
```
Comment (qiwjin) [inline src/Microsoft.Health.Fhir.Liquid.Converter/Processors/FhirProcessor.cs:10]: is it an inheritance? do we have other option?

File: data/Templates/Stu3ToR4/metadata.json (hunk 1 above)
Comment (sowu880) [reply data/Templates/Stu3ToR4/metadata.json:2]: Reasons not include version information in datatype:
1. If add version here, there will be too many types in the future. And most of these datatypes will share same logic and same engine.
2. Include version here such as "fhir-stu3", "fhir-r4" are not at the same level with previous data types ("hl7v2"and "ccda").
3. For the current metadata format, it's not sufficient to present entire version information. Based on our previous assumption, converter's output is FHIR R4 format. If r4->r5 added, we needs to re-design the format of metadata to represent both input data type and output data type. 

For these three reasons, we add "fhir" here to distinguish FHIR resources with general JSON, and this data type is the same level with "hl7v2" and "ccda". 

If we really need to distinguish different versions in the future, there are two ways:
1. Re-design metadata structure: add new sub-field to represent input and output version if data type is "fhir". As for converter's API, add new parameters for input and output version.
2. Add version information into templates, and make it transparent for converter engine.

Comment (moria97) [reply data/Templates/Stu3ToR4/metadata.json:2]: Got it. Thanks

File: src/Microsoft.Health.Fhir.Liquid.Converter.Tool/ConverterLogicHandler.cs (hunk 2 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.Liquid.Converter.Tool/ConverterLogicHandler.cs]: Thanks, removed switch

File: src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs [hunk 4]
```diff
@@ -399,6 +430,28 @@ public async Task GivenJsonSameInputData_WithDifferentTemplateSource_WhenConvert
...
+            var fhirProcessor = new FhirProcessor();
+            var inputContent = File.ReadAllText(inputFile);
+
+            var imageResult = fhirProcessor.Convert(inputContent, rootTemplate, imageTemplateProvider);
+            var folderResult = fhirProcessor.Convert(inputContent, rootTemplate, folderTemplateProvider);
+
+            //var imageResultObject = JObject.Parse(imageResult);
+            var folderResultObject = JObject.Parse(folderResult);
+
+            //Assert.True(JToken.DeepEquals(imageResultObject, folderResultObject));
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs:452]: Do we need to add these two lines back?

File: src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs [hunk 5]
```diff
@@ -208,7 +211,7 @@ public async Task GivenDefaultTemplateReference_WhenGetTemplateCollectionFromTem
             ImageInfo imageInfo = ImageInfo.CreateFromImageReference(imageReference);
             var newTemplateCollectionProvider = new TemplateCollectionProvider(imageInfo, _emptyClient, _cache, _defaultConfig);
             var templateCollection = await newTemplateCollectionProvider.GetTemplateCollectionAsync();
-            Assert.True(expectedTemplatesCounts <= templateCollection.First().Count());
+            Assert.Equal(expectedTemplatesCounts, templateCollection.First().Count());
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs:214]: In previous discussion, it looks that we don't need to assert exact equal, as templates are evolving all the time.

Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs:214]: Details could be found in this PR: https://github.com/microsoft/FHIR-Converter/pull/360/files

File: src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs (hunk 4 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs]: Thanks, fixed it.

File: data/Templates/Stu3ToR4/metadata.json (hunk 1 above)
Comment (yankunhuang-pku) [inline data/Templates/Stu3ToR4/metadata.json:2]: Is it case sensitive?

Comment (sowu880) [reply data/Templates/Stu3ToR4/metadata.json:2]: Yes, it's case sensitive.

File: src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs (hunk 5 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs]: Thanks, changed it back to "<=".

Comment (sowu880) [reply src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs]: Thanks, changed it back to "<=".

File: src/Microsoft.Health.Fhir.Liquid.Converter/Processors/FhirProcessor.cs (hunk 3 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Processors/FhirProcessor.cs:10]: After offline discussion, we added doc here for refinement Todo list.

File: src/Microsoft.Health.Fhir.Liquid.Converter/Processors/FhirProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        public FhirProcessor(ProcessorSettings processorSettings = null)
+            : base(processorSettings)
+        {
+        }
+
+        public override string Convert(string data, string rootTemplate, ITemplateProvider templateProvider, TraceInfo traceInfo = null)
+        {
+            // TO DO For FHIR Specific Logic
+            // 1. Add FHIR related checks for pre-processor and post-procesor.
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.Liquid.Converter/Processors/FhirProcessor.cs:20]: Nit: post-processor

File: src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs [hunk 7]
```diff
@@ -37,6 +37,7 @@ public class TemplateCollectionFunctionalTests : IAsyncLifetime
         private readonly string _defaultHl7v2TemplateImageReference = "microsofthealth/hl7v2templates:default";
         private readonly string _defaultCcdaTemplateImageReference = "microsofthealth/ccdatemplates:default";
         private readonly string _defaultJsonTemplateImageReference = "microsofthealth/jsontemplates:default";
+        private readonly string _defaultStu3ToR4TemplateImageReference = "microsofthealth/stu3tor4templates:default";
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs:40]: Just curious have we pushed this image already?

File: src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs [hunk 8]
```diff
@@ -85,10 +87,11 @@ public static IEnumerable<object[]> GetImageInfoForArtifactWithTag()
 
         public static IEnumerable<object[]> GetDefaultTemplatesInfo()
         {
-            yield return new object[] { "microsofthealth/fhirconverter:default", 852 };
-            yield return new object[] { "microsofthealth/hl7v2templates:default", 852 };
+            yield return new object[] { "microsofthealth/fhirconverter:default", 915 };
+            yield return new object[] { "microsofthealth/hl7v2templates:default", 915 };
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs:91]: Do we need to change these two lines?

File: src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs [hunk 9]
```diff
@@ -150,6 +151,35 @@ public static IEnumerable<object[]> GetJsonDataAndTemplateSources()
...
+                @"ValueSet",
+            };
+            return data.Select(item => new[]
+            {
+                Path.Join(_sampleDataDirectory, "Stu3", $"{item}.json"),
+                Path.Join(_templateDirectory, "Stu3ToR4"),
+                item,
+            });
+
+            return data.Select(item => new object[]
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs:175]: return after return?

Comment (sowu880) [reply src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs]: Thanks, fixed it.

File: src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs (hunk 8 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.TemplateManagement.UnitTests/Providers/TemplateCollectionProviderTests.cs:91]: After offline discussion, we can update this value regularly when templates are stable.

File: src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs (hunk 7 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs:40]: This reference represents the build-in template zip package.

Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.TemplateManagement.FunctionalTests/TemplateCollectionFunctionalTests.cs:40]: Yes, just want to make sure that we publish this default package when we release stu3 to r4 conversion.

**Review by qiwjin — APPROVED** [review]

**Review by BoyaWu10 — APPROVED** [review]

**Review by yankunhuang-pku — APPROVED** [review]

**Pull request merged**